runtime/src/accounts_index.rs

```diff
@@ -181,9 +208,198 @@ impl<'a, T: 'static + Clone> Iterator for AccountsIndexIterator<'a, T> {
+    // writing updates to the index for this Pubkey at the same time!
+    fn remove_by_inner_key(&self, inner_key: &Pubkey, slots_to_remove: Option<&HashSet<Slot>>) {
+        // Save off which keys in `self.index` had slots removed so we can remove them
+        // after we purge the reverse index
+        let mut key_to_removed_slots: HashMap<Pubkey, Vec<Slot>> = HashMap::new();
+
+        // Check if the entry for `inner_key` in the reverse index is empty
+        // and can be removed
+        let needs_remove = {
+            if let Some(slots_to_remove) = slots_to_remove {
+                self.reverse_index
+                    .get(inner_key)
+                    .map(|slots_map| {
+                        // Ideally we use a concurrent map here as well to prevent clean
+                        // from blocking writes, but memory usage of DashMap is high
+                        let mut w_slots_map = slots_map.value().write().unwrap();
+                        for slot in slots_to_remove.iter() {
+                            if let Some(removed_key) = w_slots_map.remove(slot) {
+                                key_to_removed_slots
+                                    .entry(removed_key)
+                                    .or_default()
+                                    .push(*slot);
+                            }
+                        }
+                        w_slots_map.is_empty()
+                    })
+                    .unwrap_or(false)
+            } else {
+                if let Some((_, removed_slot_map)) = self.reverse_index.remove(inner_key) {
+                    for (slot, key) in removed_slot_map.into_inner().unwrap().into_iter() {
+                        key_to_removed_slots.entry(key).or_default().push(slot);
+                    }
+                }
+                // We just removed the key, no need to remove it again
+                false
+            }
+        };
+
+        if needs_remove {
+            if let dashmap::mapref::entry::Entry::Occupied(slot_map) =
+                self.reverse_index.entry(*inner_key)
+            {
+                if slot_map.get().read().unwrap().is_empty() {
+                    slot_map.remove();
+                }
+            }
+        }
+
+        // Remove this value from those keys
+        for (key, slots) in key_to_removed_slots {
+            self.remove_index_entries(&key, inner_key, &slots);
+        }
+    }
+
+    fn get(&self, key: &Pubkey) -> Vec<Pubkey> {
+        if let Some(inner_keys_map) = self.index.get(key) {
+            inner_keys_map
+                .iter()
+                .map(|entry_ref| *entry_ref.key())
+                .collect()
+        } else {
+            vec![]
+        }
+    }
+}
+
 #[derive(Debug, Default)]
 pub struct AccountsIndex<T> {
     pub account_maps: RwLock<AccountMap<Pubkey, AccountMapEntry<T>>>,
+    program_id_index: SecondaryIndex,
+    token_owner_index: SecondaryIndex,
+    mint_index: SecondaryIndex,
```

Comment:
```suggestion
    spl_token_owner_index: SecondaryIndex,
    spl_token_mint_index: SecondaryIndex,
```

Reply:
Done in https://github.com/solana-labs/solana/pull/14212/commits/d204a08432c3936cd1450cf16e1a67236e758461